### PR TITLE
BUG: Index.str.cat casting result always to object

### PR DESCRIPTION
**Pull request by phofl**
- [ ] closes #xxxx (Replace xxxx with the GitHub issue number)
- [ ] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [ ] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [ ] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [ ] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

### CODE REVIEW

File: pandas/core/strings/accessor.py [hunk 1]
```diff
@@ -689,12 +690,15 @@ def cat(
         out: Index | Series
         if isinstance(self._orig, ABCIndex):
             # add dtype for case that result is all-NA
+            dtype = None
+            if isna(result).all():
```
Comment (mroeschke) [inline pandas/core/strings/accessor.py:694]: Does the comment in `test_str_cat_all_na` refer to the the object dtype here (othewise good to add a TODO comment here)

Comment (phofl) [reply pandas/core/strings/accessor.py:694]: Yes that's the place where I found this being an issue

**Comment by phofl** [comment]
cc @mroeschke

**Review by mroeschke — APPROVED** [review]

**Comment by mroeschke** [comment]
Thanks @phofl

**Pull request merged**